numpy/_core/function_base.py

```diff
@@ -126,21 +127,20 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
         )
     div = (num - 1) if endpoint else num
 
-    # Convert float/complex array scalars to float (gh-3504),
-    # and make sure one can use variables that have
-    # an __array_interface__ (gh-6634).
-    start = asanyarray(start) * 1.0
-    stop = asanyarray(stop) * 1.0
+    conv = _array_converter(start, stop)
+    start, stop = conv.as_arrays(subok=True)
+    dt = conv.result_type(ensure_inexact=True)
```

Comment:
Iterated a bit, this is still a bit rough (I would like to consolidate `__array_wrap__` a bit at least).

This does define `conv.wrap()` as the one true way to call `__array_wrap__`, but for NEP 50, the interesting part is really this part.
(I haved the cast to float into the subtract via `dtype=`, which isn't strictly necessary.)

The point is, that this means `np.linspace(np.float32(3), 10)` returns float32.  (yes, still needs tests, but thought I would note this, as this pattern would hopefully be useful a few places, that are currently subtly broken in the same way.)

Reply:
@mhvk beyond cleanup, part of my thought for consolidating is that it should help with gh-24897.  Because wrapping is a reasonable place to decide scalar vs. array logic.